Add List component

apedroferreira opened the pull request:
Closes https://github.com/mui/mui-toolpad/issues/1487

Allows for building a layout from the first item position in the list, which is repeated a number of times as per the `itemCount` prop. Inside each list item, the respective `i` index can be accessed in the local bindings.

Introduces templating + local scopes, and scoped props with the `isScoped` property in the arg types, which makes the prop have a unique value per scope.

https://user-images.githubusercontent.com/10789765/216412886-06026c5a-831c-46b0-8a90-92e6d12dcec2.mov


Preview: https://toolpad-pr-1527.onrender.com/

bytasv commented on packages/toolpad-components/src/List.tsx:
```diff
@@ -0,0 +1,47 @@
+
+function List({ renderItems, disablePadding = false, sx }: ListProps) {
+  return (
+    <MuiList disablePadding={disablePadding} sx={{ width: '100%', ...sx }}>
```
It might be worth to consider if we shouldn't have a generic list construct where we could use any container as a list item and then pass children inside that container so that those children elements would be rendered by iterating through list array.

I.e. once we have `CardComponent` I would expect that I can use "list" generic component to render array of items as cards.

Since this isn't what was originally specified in the task feel free to ignore my comment. I just think it's worth to have this in mind for future

apedroferreira replied:
On an unrelated note, I might add a generic `Container` component as a stripped-down version of our current `Paper` component - it follows the same kind of thinking.

bytasv replied:
`Iterator` I think it something what I'm thinking about. As I mentioned I would expect that we should be able to take any group of component and render it in array, thinking a list of Cards and how could that be done.
The only minor issue that I see with `Iterator` is that it's very technical term and `List` might be a more natural name to what it is suppose to do

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/PageEditor/RenderPanel/RenderOverlay.tsx:
```diff
@@ -824,7 +824,12 @@ export default function RenderOverlay({ bridge }: RenderOverlayProps) {
 
         // Detect center in layout containers
-        if (isDraggingOverElement && activeDropNodeInfo && activeDropSlot) {
+        if (
```
It's all bug fixes in this file.

apedroferreira commented on packages/toolpad-app/src/runtime/ToolpadApp.tsx:
```diff
@@ -191,13 +206,25 @@ function RenderedNodeContent({ node, childNodeGroups, Component }: RenderedNodeC
 
+  const { item: iteratorItem } = useIteratorItemContext();
+
+  const globalScopeNodeState = React.useMemo<GlobalScopeNodeState>(
```
Use context to get active item from where the items have been mapped.

apedroferreira commented on packages/toolpad-app/src/runtime/ToolpadApp.tsx:
```diff
@@ -353,20 +380,37 @@ function RenderedNodeContent({ node, childNodeGroups, Component }: RenderedNodeC
         );
       }
+
+      props[propName] = isElementIterator
```
Wrap element iterator slot children by iterating and creating a context for the active item.

apedroferreira commented on packages/toolpad-app/src/runtime/ToolpadApp.tsx:
```diff
@@ -671,6 +719,22 @@ function parseBindings(
         }
       }
+
+      const firstIteratorAncestorItems = getFirstElementIteratorAncestorItems(dom, elm, components);
```
Create local bindings for each node with an element iterator as parent.

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/GlobalScopeExplorer.tsx:
```diff
@@ -39,6 +39,10 @@ function groupScopeMeta(value: Record<string, unknown>, meta: GlobalScopeMeta):
       displayName: 'Queries',
       items: [],
     },
+    hidden: {
```
Introduce hidden bindings, as there are bindings created for some nodes that I didn't want to show in others.

apedroferreira commented on packages/toolpad-components/src/List.tsx:
```diff
@@ -0,0 +1,47 @@
+      defaultValue: [...Array(3).fill({})],
+    },
+    renderItems: {
+      typeDef: { type: 'elementIterator', itemsProp: 'items' },
```
Using this `itemsProp` so that the iterator is aware of what the prop for the items is. Lmk if there are any ideas to define this differently.

prakhargupta1 commented:
When a text box is dropped in a list component, just like a button should it also not appear thrice but treated as a single component so that it can handle array? In Retool it works like that.

apedroferreira commented:
> When a text box is dropped in a list component, just like a button should it also not appear thrice but treated as a single component so that it can handle array? In Retool it works like that.

Sorry, I didn't understand, could you show a video example of Retool to see what you mean maybe?
If it's about the Text component not showing 3 times when I place it in the video, it seems that is because the default value for the Text component has been removed? But it wasn't in this PR.
It probably happened unintentionally in some recent commit?

Janpot commented on packages/toolpad-app/src/canvas/getPageViewState.ts:
```diff
@@ -94,6 +94,11 @@ export function getNodesViewInfo(rootElm: HTMLElement): {
           const parentId: NodeId = fiber.memoizedProps.parentId as NodeId;
           const nodeSlots = nodes[parentId]?.slots;
 
+          // If there are 2 slots for the same prop name, consider only the first one
```
Which scenario would that be?

apedroferreira replied:
Without this exception, every item rendered by the iterator would have its own slot where you can build layout - and any changes made in any of those slots would always apply to every item. 
So this was a quick way to make it so that in cases like this it's only possible to build a layout in the first slot, using it as a template.

Janpot commented on packages/toolpad-core/src/types.ts:
```diff
@@ -75,7 +75,15 @@ export type BindableAttrEntries = [string, BindableAttrValue<any>][];
+    | 'object'
+    | 'array'
+    | 'element'
+    | 'elementIterator'
```
It wouldn't be possible to come up with a more generic construct, that allows us to expand templating to other situations? I'm thinking custom column types. Or custom autocomplete items.

apedroferreira replied:
Do you have more specific examples of limitation this approach has that wouldn't work for those use cases?

**Custom column types:** if this would be a Datagrid feature, it might deserve its own specific approach?
**Custom autocomplete items:** what this new `elementIterator` does is allow us to render a template n number of times based on the layout of the first item, and wrap those items with whatever we want to (in the case of this `List`, a `ListItem`). So it should be possible to use an `elementIterator` for custom autocomplete items?

Janpot replied:
There are two concepts here:

* The concept of "a template", which describes what this prop holds. a template that can be used to render multiple elements, based ona  common template
* The concept of "a template editing UI". This is what the "slot" control does for the `element` type. It defines how a user interacts with the property.

It seems to me that you are bundling these concepts together here in the `elementIterator` type. My expectation would be that there is a `template` type. and something like a `dragAndDrop` control. That way we separate the logic of what the prop holds from how it's being manipulated. e.g. you can have a single `number` type, that can be manipulated by several different controls.

apedroferreira replied:
But for example the `Paper` component is already also using a `layoutSlot` for the same layout building experience, even though it does not iterate, so they're already separate concepts? Maybe it's just a question of naming things differently?
We can clear up these details later synchronously though and finish the discussion about this.

bytasv reviewed (changes requested):
All in all seems to be working fine as in to render list of items, but I found a bug (I suppose?) when I delete item, but there is still one remaining it deletes both elements 👇 

https://user-images.githubusercontent.com/437214/212868101-0d95351f-dd65-4d71-8e2b-d3df02d3b9e6.mp4

apedroferreira commented:
@bytasv I've tried to make the part you mentioned more readable + fixed the bug you found with layout slots (thanks for finding it!). Let me know if it's good now.

bytasv reviewed (approved):
Works well now, great job

Janpot commented:
We had in-depth 1-on-1 around this PR. Thinking past week about the extent of this problem, I had a hunch we were underestimating the complexity. To illustrate the issues, consider following retool vs. toolpad apps:

Retool: https://mui.retool.com/editor/3cf0ba28-c1fe-11ec-95db-df6327e52470/jans%20space
Toolpad: https://toolpad-pr-1527.onrender.com/_toolpad/app/cld8ygo8g0001kh9t719qc3wk/pages/331kqzd

Main areas to note are:
1. Retool app list items create dom based on a template, but with a new binding scope, Toolpad app list items are copied dom nodes, within the global scope. Therefore the behavior differs considerably between implementation. This app is mostly to illustrate the problem very visually.
2. The input label tries to bind to the list iterator but it's not defined.


We will have to investigate around how to create nested scopes within the binding engine. Currently there is nothing foreseen for this.

apedroferreira commented:
> Retool app list items create dom based on a template, but with a new binding scope, Toolpad app list items are copied dom nodes, within the global scope. Therefore the behavior differs considerably between implementation. This app is mostly to illustrate the problem very visually.

Better defined local scopes would help this but there also seems to be a difference in how Retool handles input values compared to how we do... In Toolpad the input value comes directly from a prop in the Text Field component, but there's no `value` control in Retool, the value seems to be uncontrolled but still retrievable in bindings somehow. Any other controlled prop in Retool also just copies itself unless we set a binding...

> The input label tries to bind to the list iterator but it's not defined.

I think you misunderstood this part - there is no `i` in the current implementation.

<img width="1195" alt="Screenshot 2023-01-23 at 18 18 04" src="https://user-images.githubusercontent.com/10789765/214118313-89bff937-d55c-4ad6-85a8-6387646e0683.png">

What we have now is, in this case, this `textFieldItem` in the local bindings which retrieves what you expect to be `items[i]`. The `items` prop needs to be set and there would be an object there, as also shown in my video example above. We could also easily make an index available if it would be useful.

bytasv commented:
Even if Toolpad differs from retool implementation and does not support "local scopes" does it makes this a blocker in this case? I'd assume this would come as a further improvement once we actually need it. Right now if I understood correctly it does what I would expect it to do - access array item according to the index and renders the data

<img width="906" alt="CleanShot 2023-01-23 at 21 49 00@2x" src="https://user-images.githubusercontent.com/437214/214135583-1bfb7860-9b6d-496d-9b8c-f39f95635fdb.png">

as far as I can tell this functions as we intend in the initial scope of this ticket 🤔

bytasv commented:
> But iterative also means that a subsequent step can build on top of the previous step.

I don't see why it has to be built on previous step. More often than not when you learn something new you you may want to refactor or rewrite it something from the ground up. I don't think this is any different. With the information that we have right now and basic requirements that were defined in the original task this DOES satisfy it. If there are technical reasons not to merge it now (like exposed scopes) then we should look into how we can patch it and unblock the main functionality. 
Once we learn all about the different use-case and further requirements for list component we can even create list v2 and start deprecating list v1 if that's the way we choose.

Iteration is about learning something based on feedback and improving it accordingly.

I can easily see a value with a current implementation of this component to the end users even if it was some drawbacks or limitations. We should fix/improve/change it based on an actual use-case or user's feedback. Unless we are committing something that would be irreversible going forward, I don't see a reason for this to be blocked/not merged until we figure out how to do a new, improved version of such component.

apedroferreira commented:
> Personally, to simplify the API further, I'd also just replace items with count and only supply i in the scope. It should be enough to display lists and iterate their items.

In my latest changes I'm passing `i` and `item`. I can pass only `i` and replace `items` with `count`, sure.
I'm also using a component API like you suggested now.

> I tried adding a click handler to a button in a list (as per your screenshot), it doesn't have access to either item or i.

Seems like my latest commits didn't deploy, I need to fix the autocomplete so the CI passes... My workarounds to make the autocomplete work were causing most of the issues you perceived, but they weren't the only way to do things.

> The lack of local scope causes those listitems to show up in all the bindings. But what does it mean to bind to them?

The global scope items were there just for the autocomplete. I just removed that and they still work in the runtime. But now I have to fix the autocomplete in some way that doesn't involve using global scope.
There's actually a "local scope" now, but it's using React context. We could see how to improve it from here?
Also I think the problem with the inputs not working like in Retool is partially a different problem related to their inputs being uncontrolled, like I mentioned.

apedroferreira commented on packages/toolpad-app/src/runtime/ToolpadApp.tsx:
```diff
@@ -145,6 +153,213 @@ function useElmToolpadComponent(elm: appDom.ElementNode): ToolpadComponent {
+  };
+}
+
+function parseBindings(
```
I just moved a few functions above in this file to be able to use them anywhere, and replaced the `page: appDom.PageNode` argument with `rootNode: appDom.ElementNode | appDom.PageNode`, but I actually ended up not using this so far, not sure if it's needed...

apedroferreira commented on packages/toolpad-components/src/TextField.tsx:
```diff
@@ -35,6 +35,7 @@ export default createComponent(TextField, {
       onChangeProp: 'onChange',
       defaultValue: '',
       defaultValueProp: 'defaultValue',
+      isScoped: true,
```
I introduced the concept of scoped parameters as in Retool input values seem unique per scope but other props are not.

Janpot commented:
This is coming along nice. I still have this problem:

* Open https://toolpad-pr-1527.onrender.com/_toolpad/app/cle309iyl0004pt9aqmx1g7h2/pages/z239t9x
* It shows only 3 items for me
* Drag a component in the canvas
* Suddenly it shows the correct 20 items

apedroferreira commented:
> It shows only 3 items for me

There was an issue with prop values that changed by themselves (e.g: query results), I've added a fix.

> If it isn't too much work I'd also add a test that verifies the different binding use-cases we discussed on Slack.

I've added a more specific test for the List component that should cover those cases.

Let me know if you find anything else!
